src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs

```diff
@@ -323,22 +230,25 @@ private static void ExpirationTokensExpired(object obj)
             {
                 var entry = (CacheEntry)state;
                 entry.SetExpired(EvictionReason.TokenExpired);
-                entry._notifyCacheOfExpiration(entry);
+                entry._cache.EntryExpired(entry);
             }, obj, CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
         }
 
         private void DetachTokens()
         {
-            lock (_lock)
+            if (_expirationTokenRegistrations != null)
```

Comment:
Isn't this a behavior change?

If another thread is already in `AttachTokens` and has the lock, `DetachTokens` would wait for the other thread to complete, and then check for `_expirationTokenRegistrations != null`.

Reply:
@eerhardt great catch! I've fixed that and added a comment that explains it